### ISSUE TITLE
Features: DTLS 1.3 and QUIC

### ISSUE THREAD

**Issue by seppy87** [randombit/botan#1961]
Hallo,

since my DTLS Server and Client are working now 👍 I am wondering what is planned for the future in this library?

DTLS v1.3 ?
QUIC ?

and since I am using dynamic bindings. Do you wish to have a special thanks/mentioning for your project in my Documentation?

Greetings

**Comment by randombit** [comment]
TLS 1.3 support is on the horizon and DTLS 1.3 support will presumably follow. IIRC the DTLS 1.3 RFC hasn't been finalized yet. Either this year or next year.

QUIC is (IIRC, and I haven't read the recent drafts) based on/using TLS 1.3 handshake, so naturally it would come behind 1.3 support in any case. An full QUIC impl in the library is somewhat less likely, unless someone stepped up who had an interest in driving and maintaining the code. (Read: I don't have a personal need for QUIC nor has anyone offered to pay me to write it, so for me it falls behind improvements that meet one of those criteria). More likely endpoint is we just offer TLS 1.3 with the necessary hooks and primitives needed to implement QUIC in an external library.

**Comment by seppy87** [comment]
Probably the amount you are thinking of exceeds the amount I am capable of as normal human.

**Comment by reneme** [comment]
It turns out we have a use case for DTLS 1.3 and will likely start working on an implementation based on the existing TLS 1.3 code. I did a first read-through of [RFC 9147](https://www.rfc-editor.org/rfc/rfc9147.html) and below are some scratchy notes on what probably has to be done. At this point, we're not ready to provide an exact time frame for this, but it should be measured in months not in years.

**TL;DR:** We'll probably need alternative implementations or significant extensions of `TLS::Record_Layer`, `TLS::Handshake_Layer`, and `TLS::Cipher_State`. All the higher layers of the protocol, particularly the implementation of the actual handshake messages and the TLS state machines in `TLS::[Client/Server]_Impl_13` can probably stay mostly untouched.

* **Record Layer**
  * Support demux of DTLS records
  * Handling of Epoch and Sequence Number
  * Deprotection of Epoch/SeqNumber (see [here](https://www.rfc-editor.org/rfc/rfc9147.html#name-record-number-encryption))
  * Replay protection via a sliding window per epoch

* **Handshake Layer**
  * Fragmentation (additional header values + handling of them)
    Note: Transcript hash is calculated from the TLS marshalling (w/o the additional headers)
  * Timeout and Retransmission
    * Handle the retransmission state machine (see [RFC 9147 Section 5.8.1](https://www.rfc-editor.org/rfc/rfc9147.html#section-5.8.1))
    * Example: https://www.rfc-editor.org/rfc/rfc9147.html#section-6
    * Probably `Handshake_Layer_13` will need to be able to actively create messages (i.e. ACKs)

* **Key Management in `Cipher_State`**
  * Explicit cipher state handling (Epochs)
    Out-of-order messages at epoch boundaries requires keeping multiple sets of keys for some period
  * Adopt DTLS-specific HKDF labels

* **Handshake Messages and Extensions**
  * ClientHello/HelloRetryRequest
    * Add active support of [the Cookie extension](https://www.rfc-editor.org/rfc/rfc8446#section-4.2.2) for [DoS protection](https://www.rfc-editor.org/rfc/rfc9147.html#name-denial-of-service-counterme) in the server implementation.
    * Optionally always enforce Hello Retry Request
  * Client Hello
    * additionally contains a legacy_cookie field from DTLS 1.2
  * ACK Message

  * Optional? New Post-Handshake Message: RequestConnectionID
  * Optional? New Post-Handshake Message: NewConnectionID
  * Optional? [Connection_ID extension](https://www.rfc-editor.org/rfc/rfc9146.html)

* **Misc**
  * DTLS 1.3 _does not_ support TLS 1.3 "compatibility mode"

* **Additional User API**
  * Callbacks to establish and validate ClientHello Cookies
  * Configuration of estimated Path MTU?
  * Retransmission trigger (avoid having timers within Botan)
